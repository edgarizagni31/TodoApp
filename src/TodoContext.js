import { createContext, useEffect, useReducer, useState } from 'react';
import { todoReducer } from './helpers/todoReducer';

export const TodoContext = createContext();

const  TodoProvider = ( {children} ) => {
    const init = () => JSON.parse(localStorage.getItem('todos')) || [];
    const [ todos, dispatch ] = useReducer( todoReducer, [], init );
    const [ listTodos, setListTodos ] = useState(todos);

    useEffect( 
        () => {
            localStorage.setItem('todos', JSON.stringify(todos));
            setListTodos( todos );
        }
    , [ todos ])
   
    const filterComplete =   () => {
      setListTodos( todos.filter( (t) => t.done === true ) );
    }
  
    const filterActives =  () => {
      setListTodos( todos.filter( (t) => t.done === false ) )
    }

    const resetTodos = () => {
      setListTodos( todos );
    }

    return (
        <TodoContext.Provider 
            value = {
                {
                    listTodos,
                    dispatch,
                    resetTodos,
                    filterActives,
                    filterComplete
                }
            }
        >
            {children}
        </TodoContext.Provider>
    )
}

export default TodoProvider;
