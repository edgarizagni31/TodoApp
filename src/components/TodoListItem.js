import React from "react";

import { BsX } from "react-icons/bs";

export const TodoListItem = ({ todo, deleteTodo, toogleTodo }) => {
  const handleRemoveTodo = (id) => deleteTodo(id);
  const handleToogleTodo = (id) => toogleTodo(id);

  return (
    <div className='todo-list__item'>
      <button
        className={`${todo.done && "todo-list__btn--active"} todo-list__btn `}
        onClick={() => handleToogleTodo(todo.id)}
      ></button>
      <p className={`${todo.done && 'complete'} todo-list__text`}>
        {todo.desc}
      </p>
      <BsX
        className='todo-list__icon'
        onClick={() => handleRemoveTodo(todo.id)}
      />
    </div>
  );
};
