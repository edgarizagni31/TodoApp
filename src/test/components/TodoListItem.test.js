import { mount } from 'enzyme';
import { TodoContext } from '../../TodoContext';
import TodoListItem from '../../components/TodoListItem';
import { demoTodos } from '../fixtures/demoTodo';

describe('test to TodoListItem', () => {
    const dispatch = jest.fn();
    const todo = demoTodos[0];
    const wrapper = mount(
        <TodoContext.Provider  value = { { dispatch } }>
           <TodoListItem todo = { todo }/>
        </TodoContext.Provider>
    );

    test('should be displayed correctly', () => {
        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('button').length ).toBe(1);
        expect( wrapper.find('p').text().trim() ).toBe( todo.desc )
        expect( wrapper.find('BsX').length ).toBe(1);
    });

    test('should mark the todo as complete', () => {
        wrapper.find('button').simulate('click', todo.id );
        expect( dispatch ).toHaveBeenCalledTimes(1);
        expect( dispatch ).toHaveBeenCalledWith({ type: 'toggle', payload: todo.id })
    });

    test('should mark the todo as delete', () => {
        wrapper.find('BsX').simulate('click', todo.id );
        expect( dispatch ).toHaveBeenCalledTimes(1);
        expect( dispatch ).toHaveBeenCalledWith({ type: 'delete', payload: todo.id })
    });
})
