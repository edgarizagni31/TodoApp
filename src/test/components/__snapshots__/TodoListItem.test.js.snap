// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test to TodoListItem should be displayed correctly 1`] = `
<div
  className="todo-list__item"
>
  <button
    aria-label="completado"
    className="false todo-list__btn "
    onClick={[Function]}
  />
  <p
    className="false todo-list__text"
  >
    Aprender mongo
  </p>
  <svg
    aria-label="eliminar todo"
    className="todo-list__icon"
    fill="currentColor"
    height="1em"
    onClick={[Function]}
    stroke="currentColor"
    strokeWidth="0"
    style={
      Object {
        "color": undefined,
      }
    }
    viewBox="0 0 16 16"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      clipRule="evenodd"
      d="M11.854 4.146a.5.5 0 010 .708l-7 7a.5.5 0 01-.708-.708l7-7a.5.5 0 01.708 0z"
      fillRule="evenodd"
      key="0"
    />
    <path
      clipRule="evenodd"
      d="M4.146 4.146a.5.5 0 000 .708l7 7a.5.5 0 00.708-.708l-7-7a.5.5 0 00-.708 0z"
      fillRule="evenodd"
      key="1"
    />
  </svg>
</div>
`;
